src/pybind/mgr/dashboard/controllers/osd.py

```diff
@@ -229,6 +232,33 @@ def safe_to_destroy(self, ids):
                 'is_safe_to_destroy': False,
             }
 
+    @Endpoint('GET', query_params=['ids'])
+    @ReadPermission
+    def safe_to_remove(self, ids):
+        """
+        :type ids: int|[int]
+        """
+        # Unused now, might be useful for individual OSD checks in the future
+        _ = ids
+
+        # Check if there are any OSDs exceeds nearfull or full threshold.
+        all_health = health.HealthData(self._has_permissions, minimal=True).all_health()
+        checks = {c['type'] for c in all_health['health']['checks']}
+        unsafe_checks = set(['OSD_FULL', 'OSD_BACKFILLFULL', 'OSD_NEARFULL'])
+        failed_checks = checks & unsafe_checks
+        return {
+            'is_safe_to_remove': not bool(failed_checks),
+            'message': 'Removing OSD(s) is not recommended with these health checks: {}.'.
+                       format(', '.join(failed_checks)) if failed_checks else ''
+        }
```

Comment:
The issues I see with this one are:
* How we can state, in advance, that removing more than 1 OSD at the same time is safe, based on these checks (or even in other ones). I'd only dare doing so if one by one, and after any data migration has finished (maybe I'm missing something here: is every orchestrator taking care of the graceful removal of OSDs?).
* If I'm not wrong, all Ceph checks come from core Ceph as a single source of Truth, so I think in any case we should try to move this (or any new health check) to the core.
* Additionally, as there's no way to ensure transaction-safety here, there could be the chance of cluster status changing right after this check succeeded and before the removal of OSDs (or even in between).
* With a CLI, if you make a mistake, you may try to quickly press CTRL+C and pray, but this lies on top of a really complex engine almost impossible to stop when triggered.

With operations that are destructive, irreversible and could also have integrity & operational impact, I always try make them as stiff & slow as the most patient user could tolerate. Another approach that usually grants operators a better sleep time is to count on a 2-stage deletion process (aka Trash/Recycle bin), so there's time to react/see how Ceph reacts/see how Ceph clients react.

Reply:
Thanks for the comment!
> The issues I see with this one are:
> 
> * How we can state, in advance, that removing more than 1 OSD at the same time is safe, based on these checks (or even in other ones). I'd only dare doing so if one by one, and after any data migration has finished (maybe I'm missing something here: is every orchestrator taking care of the graceful removal of OSDs?).


> * If I'm not wrong, all Ceph checks come from core Ceph as a single source of Truth, so I think in any case we should try to move this (or any new health check) to the core.
> * Additionally, as there's no way to ensure transaction-safety here, there could be the chance of cluster status changing right after this check succeeded and before the removal of OSDs (or even in between).

I agree with you that removing one OSD at a time is much safer. If the API supports removing multiple OSDs at one time, the orchetrator backend should check if next OSD removing is fesiable.

It would be nice if an orchestrator backend can have a call to return if removing certain OSDs are safe. If there is not, we can only perform some general checks.

How many operations the orchestrator backend will perform in `remove_osds` call still needs to be discussed. For now, ansible orchestrator out/purge the OSDs and zap the disks. ssh-orchestrator simply brings down the OSD container. None of them takes care of data migration.

My implementation here is just assuming that `remove_osds` call will gracefully handle all the stuff. Which is not true for now.

> * With a CLI, if you make a mistake, you may try to quickly press CTRL+C and pray, but this lies on top of a really complex engine almost impossible to stop when triggered.
> 
> With operations that are destructive, irreversible and could also have integrity & operational impact, I always try make them as stiff & slow as the most patient user could tolerate. Another approach that usually grants operators a better sleep time is to count on a 2-stage deletion process (aka Trash/Recycle bin), so there's time to react/see how Ceph reacts/see how Ceph clients react.

Not sure if this is needed. Generally, when users need to remove OSDs, they might really need to do it or already evaluate the situation carefully. For example, if they want to replace nodes: they would create OSDs with new nodes then remove some old nodes or vice versa.

Reply:
You can call https://ceph-command-api.readthedocs.io/en/latest/mon_command_api.html#ceph_command_api.MonCommandApi.osd_safe_to_destroy of course.

Reply:
I didn't use that because I thought `remove_osd` contains operations like `out` and `destroy` before service remove. If the user needs to take care of data durability by himself before removing OSDs, then this check should be sufficient.

Reply:
Yeah, that's my point @bk201, with this multilayered approach we are sometimes missing who's 'accountable' and hence responsible for putting the security measures in place. We are partly doing that in our dashboard front-end, but perhaps blindly assuming that inner layers are taking their precautions and ensuring operations come to a good end (while they're mostly executors of outer layers' commands).

Reply:
@sebastian-philipp 

The check is moved into orchestrator base class and is performed in CLI: (https://github.com/ceph/ceph/pull/31997/commits/b78bb82c676b57973a89a660c1141cc72c27855a)